cmd/loadtest/loadtest.go

```diff
@@ -0,0 +1,357 @@
+				}
+			}
+		}(ctx, l)
+	}
+
+	const samples = 25
+	for {
+		var totalDuration time.Duration
+		for range samples {
+			select {
+			case <-ctx.Done():
+				return
+			case l := <-updateLatencyChan:
+				klog.V(1).Infof("Single update: %s", l)
+				totalDuration += l
+			}
+		}
+		avgDuration := totalDuration / samples
+		klog.Infof("Average update time is %s with throttle %s", avgDuration, thr)
+		if avgDuration < *timeout && thr.oversupply == 0 {
+			if thr.opsPerSecond > *successQPS {
+				klog.Infof("Exiting successfully: QPS reached %d", thr.opsPerSecond)
+				os.Exit(0)
+			}
+			thr.increase()
+		}
+		if avgDuration > *timeout {
+			klog.Exitf("Witness took %s on average to handle last %d requests with throttle: %s", avgDuration, samples, thr)
+		}
+	}
+}
+
+func newInMemoryLogs(n int) inMemoryLogs {
+	logs := make([]*inMemoryLog, n)
+	for i := range n {
+		logs[i] = newInMemoryLog(i)
+	}
+	return inMemoryLogs{
+		logs: logs,
+	}
+}
+
+type inMemoryLogs struct {
+	logs []*inMemoryLog
+}
+
+func (ls inMemoryLogs) config() string {
+	s := "- Logs\n"
+	for _, l := range ls.logs {
+		s += l.config()
+	}
+	return s
+}
+
+// newInMemoryLog creates a new in memory log that is initially empty. Given a particular
+// seed, the log will act deterministically in all parts of its behaviour, including signing
+// key generation, and leaf contents.
+func newInMemoryLog(seed int) *inMemoryLog {
+	origin := fmt.Sprintf("example.com/inmemorylog/%d", seed)
+
+	var cha8seed [32]byte
+	cha8seed[0] = byte(seed)
+	cha8 := rand.NewChaCha8(cha8seed)
+
+	skey, vkey, err := note.GenerateKey(cha8, origin)
+	if err != nil {
+		klog.Exitf("Failed to generate keys: %v", err)
+	}
+	s, err := note.NewSigner(skey)
+	if err != nil {
+		klog.Exitf("Failed to generate signer: %v", err)
+	}
+	genLeaf := func(i uint64) []byte {
+		return []byte(fmt.Sprintf("log %d, leaf %d", seed, i))
+	}
+	return &inMemoryLog{
+		o:       origin,
+		s:       s,
+		vkey:    vkey,
+		genLeaf: genLeaf,
+	}
+}
+
+// inMemoryLog determinstically generates log contents using genLeaf
+// and stores all of the hashes in memory. This may need rewriting as
+// memory usage is likely to be a concern for large logs.
+//
+// Anything using the hashes field was taken from
+// https://github.com/transparency-dev/merkle/blob/main/testonly/tree.go
+type inMemoryLog struct {
+	o       string
+	s       note.Signer
+	vkey    string
+	genLeaf func(uint64) []byte
+
+	mu     sync.Mutex
+	size   uint64
+	hashes [][][]byte // Node hashes, indexed by node (level, index).
+}
+
+func (l *inMemoryLog) appendLocked(hash []byte) {
+	level := 0
+	for ; (l.size>>level)&1 == 1; level++ {
+		row := append(l.hashes[level], hash)
+		hash = rfc6962.DefaultHasher.HashChildren(row[len(row)-2], hash)
+		l.hashes[level] = row
+	}
+	if level > len(l.hashes) {
+		panic("gap in tree appends")
+	} else if level == len(l.hashes) {
+		l.hashes = append(l.hashes, nil)
+	}
+
+	l.hashes[level] = append(l.hashes[level], hash)
+	l.size++
+}
```

Comment:
You can use a `compact.Range` rather than doing it all from first principles - should save you 25+ lines of quite complex code:

```go
type inMemoryLog struct {
        ... 
  	state *compact.Range  // = (&compact.RangeFactory{Hash: rfc6962.DefaultHasher.HashChildren}).NewEmptyRange(0)
	store map[string][]byte
}

func (l *inMemoryLog) appendLocked(hash []byte) {
	l.state.Append(hash, func(id compact.NodeID, h []byte) {
		l.store[fmt.Sprintf("%x/%x", id.Level, id.Index)] = h
	})
}

// Hash returns the current root hash of the tree.
func (l *inMemoryLog) Hash() []byte {
	r, _ := l.state.GetRootHash(nil)
        return r
}

func (l *inMemoryLog) getNodes(ids []compact.NodeID) [][]byte {
	hashes := make([][]byte, len(ids))
	for i, id := range ids {
		hashes[i] = l.store[fmt.Sprintf("%x/%x", id.Level, id.Index)]
	}
	return hashes
}
```

Fuller version: https://go.dev/play/p/YvSHh0cpNSp

Reply:
Done.